boltons/pathutils.py

```diff
@@ -0,0 +1,212 @@
+"""
+Functions for working with filesystem paths.
+
+The ``expandpath`` function expands the tilde to $HOME and environment variables
+to their values.
+
+The ``augpath`` function creates variants of an existing path without having to
+spend multiple lines of code spliting it up and stitching it back together.
+
+The ``shrinkuser`` function replaces your home directory with a tilde.
+
+The ``userhome`` function reports the home directory of the current user of the
+operating system.
+"""
+from __future__ import print_function
+
+from os.path import dirname
+from os.path import exists
+from os.path import expanduser
+from os.path import expandvars
+from os.path import join
+from os.path import normpath
+from os.path import split
+from os.path import splitext
+import os
+import sys
+
+
+__all__ = [
+    'augpath', 'shrinkuser', 'userhome', 'expandpath',
+]
+
+
+def augpath(path, suffix='', prefix='', ext=None, base=None, dpath=None,
+            multidot=False):
+    """
+    Create a new path with a different extension, basename, directory, prefix,
+    and/or suffix.
+
+    A prefix is inserted before the basename. A suffix is inserted
+    between the basename and the extension. The basename and extension can be
+    replaced with a new one. Essentially a path is broken down into components
+    (dpath, base, ext), and then recombined as (dpath, prefix, base, suffix,
+    ext) after replacing any specified component.
+
+    Args:
+        path (PathLike): a path to augment
```

Comment:
when we say PathLike, it means a path-like str, right?

Reply:
I'm not sure what the best "type" to use here is. What I want to communicate is that path could be either a str or pathlib.Path object.

I ultimately started using PathLike in reference this: https://docs.python.org/3/library/os.html#os.PathLike but I'm not sure if its correct. Perhaps the proper type is actually `str | PathLike`?

Reply:
I ended up swapping the input args for `str | PathLike`, but I noticed that even when a `pathlib.Path` object is the argument the output type is always `str`, so I swapped the return types for that.